fix(transformer): remove an `AstBuilder::copy` call for TS `AssignmentTarget` transform

overlookmotel opened the pull request:
Add `AssignmentTarget::get_expression_mut` method and use it to remove an unsound `AstBuilder::copy` call from TS transform for `AssignmentTarget`s.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4984?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4984** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4984?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#4983** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4983?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

DonIsaac commented on crates/oxc_transformer/src/typescript/annotations.rs:
```diff
@@ -246,10 +246,14 @@ impl<'a> TypeScriptAnnotations<'a> {
+            let inner_expr = expr.get_inner_expression_mut();
+            if inner_expr.is_member_expression() {
+                let inner_expr = self.ctx.ast.move_expression(inner_expr);
+                let Ok(member_expr) = MemberExpression::try_from(inner_expr) else {
```
Maybe unwrap instead of using `unreachable!`?

DonIsaac replied:
Or at least provide an error message in some way in case this panics

overlookmotel replied:
I used `unreachable!()` because it *is* unreachable.

I know that `unreachable!()` can cause the compiler to backtrack further to try to prove that the code is indeed unreachable, and if it can prove that, it removes the unreachable branch entirely. What I don't know is whether it does the same for `.unwrap()`. Quite possibly it does, but I don't know, so I thought better to play it safe. There is some method in my madness!

To be honest, this whole pattern of using `move_expression` and testing the same thing twice is pretty disgusting. I just don't have a good answer for how to do it better without incurring the wrath of the borrow-checker. But it's certainly better than the UB which `ast.copy` brings into play.

overlookmotel committed:
fix(transformer): remove an `AstBuilder::copy` call for TS `AssignmentTarget` transform (#4984)

Add `AssignmentTarget::get_expression_mut` method and use it to remove an unsound `AstBuilder::copy` call from TS transform for `AssignmentTarget`s.
